cmd/kubeadm/app/util/etcd/etcd.go

```diff
@@ -427,6 +453,28 @@ func (c *Client) AddMember(name string, peerAddrs string) ([]Member, error) {
 	return ret, nil
 }
 
+func memberPromoteWithRety(ctx context.Context, cli *clientv3.Client, learnerID uint64) error {
+	klog.V(1).Infof("[etcd] Promoting a learner as a voting member: %016x", learnerID)
+	// The interval and timeout are following how talos does. See:
+	// https://github.com/siderolabs/talos/blob/474604cd279def7a6798e24ede27feef955ba5a3/internal/app/machined/pkg/system/services/etcd.go#L598-L599
+	err := wait.PollImmediate(15*time.Second, 10*time.Minute, func() (bool, error) {
+		// TODO: warning logs from etcd client should be removed.
+		// The warning logs are printed by etcd client code for several reasons, including
+		// 1. can not promote yet(no synced)
+		// 2. context deadline exceeded
+		// 3. peer URLs already exists
+		// Once the client provides a way to check if the etcd learner is ready to promote, the retry logic can be revisited.
+		_, err := cli.MemberPromote(ctx, learnerID)
+		if err == nil {
+			klog.V(1).Infof("[etcd] The learner was promoted as a voting member: %016x", learnerID)
+			return true, nil
+		}
+		klog.V(4).Infof("[etcd] Promoting the learner %016x failed: %v", learnerID, err)
+		return false, err
```

Comment:
my thoughts on the return `err` instead of `nil` is that there is a looping check `wait.ExponentialBackoff` outside of this block `wait.PollImmediate`, return nil might take too long time to wait for a abnormal result, it could be 10*mins * 18.

but return `err` would potentially make the looping check here `PollImmediate` meaningless. I don't understand the logic behind `MemberPromote`, how long it would take here? so you judge this by yourself.

Reply:
`MemberPromote` won't take too much time. `A learner became ready to promote` took some time.